plugins/glustershd/rest.go

```diff
@@ -187,3 +190,280 @@ func glustershDisableHandler(w http.ResponseWriter, r *http.Request) {
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+
+	var options []string
+
+	glusterdSockpath := path.Join(config.GetString("rundir"), "glusterd2.socket")
+	options = append(options, option, "xml", "glusterd-sock", glusterdSockpath)
+	if err := txn.Ctx.Set("option", options); err != nil {
+		logger.WithError(err).Error("failed to set volname in transaction context")
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+
+	err = txn.Do()
+	if err != nil {
+		logger.WithFields(log.Fields{
+			"error":   err.Error(),
+			"volname": volname,
+		}).Error("failed to retrieve heal info")
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+	var out string
+	err = txn.Ctx.GetNodeResult(gdctx.MyUUID, "stdout", &out)
+	if err != nil {
+		logger.WithFields(log.Fields{
+			"error": err.Error(),
+		}).Error("Could Not Fetch Stdout", err)
+		return
+	}
+	output := []byte(out)
+
+	var info HealInfo
+	err = xml.Unmarshal(output, &info)
+	if err != nil {
+
+		logger.WithFields(log.Fields{
+			"error": err.Error(),
+		}).Error("Error unmarshalling from XML", err)
+		return
+	}
+
+	restutils.SendHTTPResponse(ctx, w, http.StatusOK, &info.Bricks)
+
+}
+
+func granularHealEnableHandler(w http.ResponseWriter, r *http.Request) {
+	p := mux.Vars(r)
+	volname := p["name"]
+
+	ctx := r.Context()
+	logger := gdctx.GetReqLogger(ctx)
+
+	//validate volume name
+	v, err := volume.GetVolume(volname)
+	if err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusNotFound, err)
+		return
+	}
+
+	// Store volinfo before changing Granular Heal Entry.
+	tmp := *v
+	oldvolinfo := &tmp
+
+	// validate volume type
+	if !isVolReplicate(v.Type) {
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, "Volume Type not supported")
+		return
+	}
+
+	if v.State != volume.VolStarted {
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, "Volume should be in started state.")
+		return
+	}
+
+	txn := transaction.NewTxn(ctx)
+	defer txn.Cleanup()
+
+	//Lock on Volume Name
+	lock, unlock, err := transaction.CreateLockSteps(volname)
+	if err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+
+		return
+	}
+
+	v.GranularHealEntryEnabled = true
+
+	txn.Nodes = v.Nodes()
+	txn.Steps = []*transaction.Step{
+		lock,
+		{
+			DoFunc:   "granular-heal.Enable",
+			Nodes:    []uuid.UUID{gdctx.MyUUID},
+			UndoFunc: "granular-heal.Undo",
+		},
+		{
+			DoFunc: "vol-option.UpdateVolinfo",
+			Nodes:  txn.Nodes,
+		},
+		unlock,
+	}
+
+	if err := txn.Ctx.Set("volinfo", v); err != nil {
+		logger.WithError(err).Error("failed to set volinfo in transaction context")
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+
+	if err := txn.Ctx.Set("oldvolinfo", oldvolinfo); err != nil {
+		logger.WithError(err).Error("failed to set oldvolinfo in transaction context")
```

Comment:
Use a generic error message and use log fields to add into about the keys.

Reply:
Done.